# Getting started

In [1]:
import warnings
warnings.filterwarnings('ignore')

import pandas as pd
pd.set_option('display.max_columns', None)
import numpy as np

import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
sns.set()

In [2]:
path1 = "/kaggle/input/playground-series-s3e18/"
path2 = "/kaggle/input/multi-label-classification-of-enzyme-substrates/"

sample = pd.read_csv(path1+"sample_submission.csv")
train = pd.read_csv(path1+"train.csv",index_col='id')
test = pd.read_csv(path1+"test.csv",index_col='id')
original = pd.read_csv(path2+"original.csv",index_col='id')

# Combine playground dataset and synthetic dataset generated
train1 = pd.concat([train,original]).reset_index(drop=True).drop_duplicates()

id_test = test.index # id column required for submission file
target1 = 'EC1'
target2 = 'EC2'

# Basic data exploration

In [3]:
display(train.shape),display(test.shape),display(original.shape),display(train1.shape)
print("\n")
display(train.head(3)), display(test.head(3)),display(original.head(3)),display(train1.head(3))
print("\n")
display(train.describe().applymap("{:.2f}".format)), display(test.describe().applymap("{:.2f}".format)),display(original.describe().applymap("{:.2f}".format)),display(train1.describe().applymap("{:.2f}".format))
print("\n")
train.columns, test.columns, original.columns, train1.columns

(14838, 37)

(9893, 31)

(1039, 37)

(15859, 37)

BertzCT       Chi1     Chi1n      Chi1v     Chi2n     Chi2v     Chi3v  \
id                                                                             
0   323.390782   9.879918  5.875576   5.875576  4.304757  4.304757  2.754513   
1   273.723798   7.259037  4.441467   5.834958  3.285046  4.485235  2.201375   
2   521.643822  10.911303  8.527859  11.050864  6.665291  9.519706  5.824822   

       Chi4n  EState_VSA1  EState_VSA2  ExactMolWt  FpDensityMorgan1  \
id                                                                     
0   1.749203     0.000000    11.938294  222.068080          1.181818   
1   1.289775    45.135471     0.000000  260.029719          1.346154   
2   1.770579    15.645394     6.606882  382.131027          1.085714   

    FpDensityMorgan2  FpDensityMorgan3  HallKierAlpha  HeavyAtomMolWt  \
id                                                                      
0           1.727273          2.363636          -0.24         212.163   
1           2.076923          2.769231          -0.09         247.031   
2           1.742857          2.400000          -0.78         354.106   

       Kappa3  MaxAbsEStateIndex  MinEStateIndex  NumHeteroatoms  PEOE_VSA10  \
id                                                                             
0    8.170000          11.922504        0.171585               4    0.000000   
1    3.201491          10.932338       -4.830450              10   24.415866   
2   15.033890          11.238048       -5.066255               9    0.000000   

    PEOE_VSA14  PEOE_VSA6  PEOE_VSA7  PEOE_VSA8  SMR_VSA10   SMR_VSA5  \
id                                                                      
0    91.536492        0.0   0.000000        0.0  17.744066   0.000000   
1     7.822697        0.0   0.000000        0.0   7.822697  30.705892   
2    15.645394        0.0  53.378235        0.0  15.645394  73.143616   

    SlogP_VSA3  VSA_EState9  fr_COO  fr_COO2  EC1  EC2  EC3  EC4  EC5  EC6  
id                                                                          
0     4.794537    35.527357       0        0    1    1    0    0    0    0  
1    13.825658    44.707310       0        0    0    1    1    0    0    0  
2    17.964475    45.660120       0        0    1    1    0    0    1    0

BertzCT       Chi1     Chi1n     Chi1v     Chi2n     Chi2v  \
id                                                                      
14838   344.632371   7.283603  4.473966  5.834958  3.412257  4.651530   
14839  1432.410201  10.663869  7.079026  8.065215  5.297097  5.297097   
14840    83.352608   3.931852  1.774215  1.774215  1.073446  1.073446   

          Chi3v     Chi4n  EState_VSA1  EState_VSA2  ExactMolWt  \
id                                                                
14838  2.096558  1.116433    49.458581     0.000000  204.079707   
14839  3.924155  2.569694     0.000000    29.783175  315.210331   
14840  0.467830  0.170838     5.969305     6.420822  101.047679   

       FpDensityMorgan1  FpDensityMorgan2  FpDensityMorgan3  HallKierAlpha  \
id                                                                           
14838          1.384615          1.923077          2.307692          -0.24   
14839          0.571429          0.857143          1.159286          -1.81   
14840          1.500000          2.200000          2.400000          -0.86   

       HeavyAtomMolWt     Kappa3  MaxAbsEStateIndex  MinEStateIndex  \
id                                                                    
14838         192.197   3.247123          10.248693       -4.734642   
14839         280.263  10.070000          11.917927       -0.303333   
14840          94.049   5.430000           9.694444       -0.833333   

       NumHeteroatoms  PEOE_VSA10  PEOE_VSA14  PEOE_VSA6  PEOE_VSA7  \
id                                                                    
14838               7   18.311899   13.512441   0.000000   0.000000   
14839               4    0.000000    0.000000  34.947374  98.323987   
14840               3    0.000000    5.969305   0.000000   0.000000   

       PEOE_VSA8  SMR_VSA10   SMR_VSA5  SlogP_VSA3  VSA_EState9  fr_COO  \
id                                                                        
14838   0.000000  26.809272  24.539800    4.794537    47.304082       1   
14839   9.606882   0.000000  53.378235    0.000000    43.166667       0   
14840   6.420822  11.752550  13.344559    9.589074    24.666667       1   

       fr_COO2  
id              
14838        1  
14839        0  
14840        1

BertzCT      Chi1     Chi1n     Chi1v     Chi2n     Chi2v  \
id                                                                     
C00009   49.783540  2.000000  0.782574  2.347723  0.513277  1.539831   
C00013  147.355172  3.707107  1.530297  4.590890  1.062804  3.678309   
C00014    0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   

           Chi3v     Chi4n  EState_VSA1  EState_VSA2  ExactMolWt  \
id                                                                 
C00009  0.000000  0.000000     7.822697     0.000000   97.976895   
C00013  1.914534  0.138556    15.645394     0.000000  177.943226   
C00014  0.000000  0.000000     0.000000     6.150546   17.026549   

        FpDensityMorgan1  FpDensityMorgan2  FpDensityMorgan3  HallKierAlpha  \
id                                                                            
C00009          1.200000          1.200000          1.200000           0.11   
C00013          0.888889          1.111111          1.111111           0.26   
C00014          1.000000          1.000000          1.000000          -0.04   

        HeavyAtomMolWt       Kappa3  MaxAbsEStateIndex  MinEStateIndex  \
id                                                                       
C00009          94.970  1512.242231           8.881944       -4.638889   
C00013         173.941     8.260000           9.626875       -5.050347   
C00014          14.007  -104.040000           0.000000        0.000000   

        NumHeteroatoms  PEOE_VSA10  PEOE_VSA14  PEOE_VSA6  PEOE_VSA7  \
id                                                                     
C00009               5         0.0    7.822697        0.0        0.0   
C00013               9         0.0   15.645394        0.0        0.0   
C00014               1         0.0    0.000000        0.0        0.0   

        PEOE_VSA8  SMR_VSA10  SMR_VSA5  SlogP_VSA3  VSA_EState9  fr_COO  \
id                                                                        
C00009        0.0   7.822697       0.0    4.565048    16.923611       0   
C00013        0.0  15.645394       0.0   13.440728    20.899028       0   
C00014        0.0   0.000000       0.0    0.000000     0.000000       0   

        fr_COO2  EC1  EC2  EC3  EC4  EC5  EC6  
id                                             
C00009        0    1    1    1    1    0    1  
C00013        0    1    1    1    1    0    1  
C00014        0    1    1    1    1    0    1

BertzCT       Chi1     Chi1n      Chi1v     Chi2n     Chi2v     Chi3v  \
0  323.390782   9.879918  5.875576   5.875576  4.304757  4.304757  2.754513   
1  273.723798   7.259037  4.441467   5.834958  3.285046  4.485235  2.201375   
2  521.643822  10.911303  8.527859  11.050864  6.665291  9.519706  5.824822   

      Chi4n  EState_VSA1  EState_VSA2  ExactMolWt  FpDensityMorgan1  \
0  1.749203     0.000000    11.938294  222.068080          1.181818   
1  1.289775    45.135471     0.000000  260.029719          1.346154   
2  1.770579    15.645394     6.606882  382.131027          1.085714   

   FpDensityMorgan2  FpDensityMorgan3  HallKierAlpha  HeavyAtomMolWt  \
0          1.727273          2.363636          -0.24         212.163   
1          2.076923          2.769231          -0.09         247.031   
2          1.742857          2.400000          -0.78         354.106   

      Kappa3  MaxAbsEStateIndex  MinEStateIndex  NumHeteroatoms  PEOE_VSA10  \
0   8.170000          11.922504        0.171585               4    0.000000   
1   3.201491          10.932338       -4.830450              10   24.415866   
2  15.033890          11.238048       -5.066255               9    0.000000   

   PEOE_VSA14  PEOE_VSA6  PEOE_VSA7  PEOE_VSA8  SMR_VSA10   SMR_VSA5  \
0   91.536492        0.0   0.000000        0.0  17.744066   0.000000   
1    7.822697        0.0   0.000000        0.0   7.822697  30.705892   
2   15.645394        0.0  53.378235        0.0  15.645394  73.143616   

   SlogP_VSA3  VSA_EState9  fr_COO  fr_COO2  EC1  EC2  EC3  EC4  EC5  EC6  
0    4.794537    35.527357       0        0    1    1    0    0    0    0  
1   13.825658    44.707310       0        0    0    1    1    0    0    0  
2   17.964475    45.660120       0        0    1    1    0    0    1    0

BertzCT      Chi1     Chi1n     Chi1v     Chi2n     Chi2v     Chi3v  \
count  14838.00  14838.00  14838.00  14838.00  14838.00  14838.00  14838.00   
mean     515.15      9.14      5.85      6.74      4.43      5.25      3.42   
std      542.46      6.82      4.65      5.87      3.76      4.93      3.44   
min        0.00      0.00      0.00      0.00      0.00      0.00      0.00   
25%      149.10      4.68      2.84      2.93      1.95      2.03      1.16   
50%      290.99      6.49      4.05      4.39      2.97      3.24      1.95   
75%      652.65     11.17      7.49      8.53      5.79      6.61      4.50   
max     4069.96     69.55     50.17     53.43     32.20     34.58     22.88   

          Chi4n EState_VSA1 EState_VSA2 ExactMolWt FpDensityMorgan1  \
count  14838.00    14838.00    14838.00   14838.00         14838.00   
mean       1.77       29.20       10.44     292.62             1.24   
std        1.87       31.73       13.65     225.38             5.49   
min        0.00        0.00        0.00       1.01          -666.00   
25%        0.50        5.97        0.00     148.04             1.05   
50%        1.07       17.35        6.42     206.04             1.25   
75%        2.53       44.88       12.84     343.09             1.50   
max       16.07      363.71       99.94    2237.32             3.00   

      FpDensityMorgan2 FpDensityMorgan3 HallKierAlpha HeavyAtomMolWt  \
count         14838.00         14838.00      14838.00       14838.00   
mean              1.81             2.26         -1.21         274.95   
std               5.50             5.50          0.94         212.68   
min            -666.00          -666.00         -7.73           0.00   
25%               1.69             2.10         -1.66         136.11   
50%               1.87             2.36         -1.10         194.28   
75%               2.06             2.50         -0.57         326.00   
max               3.20             3.40          0.82        2035.13   

         Kappa3 MaxAbsEStateIndex MinEStateIndex NumHeteroatoms PEOE_VSA10  \
count  14838.00          14838.00       14838.00       14838.00   14838.00   
mean       5.87             10.56          -2.12           8.58      11.02   
std       45.73              1.56           2.07           7.64      13.96   
min     -104.04              0.00          -6.33           0.00       0.00   
25%        1.78              9.93          -4.66           4.00       0.00   
50%        3.26             10.42          -1.27           6.00       6.04   
75%        5.85             11.54          -0.79          10.00      18.31   
max     1512.24             15.63           6.00          42.00      97.66   

      PEOE_VSA14 PEOE_VSA6 PEOE_VSA7 PEOE_VSA8 SMR_VSA10  SMR_VSA5 SlogP_VSA3  \
count   14838.00  14838.00  14838.00  14838.00  14838.00  14838.00   14838.00   
mean       17.79      8.96     11.32      6.70     15.67     31.07      13.64   
std        34.56     19.76     20.17     10.87     18.08     33.90      14.60   
min         0.00      0.00      0.00      0.00      0.00      0.00       0.00   
25%         0.00      0.00      0.00      0.00      5.97      6.42       4.79   
50%         5.97      0.00      0.00      0.00     11.75     20.08       9.59   
75%        15.65     12.13     13.85      6.92     17.72     42.73      14.91   
max       482.43    375.43    211.50    100.35     80.74    492.73     115.41   

      VSA_EState9    fr_COO   fr_COO2       EC1       EC2       EC3       EC4  \
count    14838.00  14838.00  14838.00  14838.00  14838.00  14838.00  14838.00   
mean        49.31      0.46      0.46      0.67      0.80      0.31      0.28   
std         29.17      0.67      0.67      0.47      0.40      0.46      0.45   
min         -5.43      0.00      0.00      0.00      0.00      0.00      0.00   
25%         30.00      0.00      0.00      0.00      1.00      0.00      0.00   
50%         41.67      0.00      0.00      1.00      1.00      0.00      0.00   
75%    

BertzCT     Chi1    Chi1n    Chi1v    Chi2n    Chi2v    Chi3v    Chi4n  \
count  9893.00  9893.00  9893.00  9893.00  9893.00  9893.00  9893.00  9893.00   
mean    516.41     9.11     5.85     6.73     4.43     5.25     3.40     1.75   
std     544.33     6.75     4.64     5.86     3.77     4.94     3.43     1.84   
min       0.00     0.00     0.00     0.00     0.00     0.00     0.00     0.00   
25%     150.26     4.70     2.85     2.93     1.95     2.05     1.17     0.51   
50%     289.90     6.45     4.01     4.34     2.93     3.17     1.92     1.06   
75%     652.76    10.97     7.49     8.53     5.79     6.52     4.30     2.51   
max    3919.54    53.88    43.28    53.99    36.37    44.73    26.74    15.62   

      EState_VSA1 EState_VSA2 ExactMolWt FpDensityMorgan1 FpDensityMorgan2  \
count     9893.00     9893.00    9893.00          9893.00          9893.00   
mean        28.96       10.53     292.01             1.28             1.86   
std         31.47       13.77     224.67             0.36             0.35   
min          0.00        0.00      15.01            -1.13             0.42   
25%          5.97        0.00     148.07             1.04             1.69   
50%         17.28        6.42     206.02             1.25             1.87   
75%         44.88       12.84     342.12             1.50             2.07   
max        363.71       99.94    1888.79             3.00             3.20   

      FpDensityMorgan3 HallKierAlpha HeavyAtomMolWt   Kappa3  \
count          9893.00       9893.00        9893.00  9893.00   
mean              2.30         -1.21         274.58     5.28   
std               0.38          0.92         212.32    37.35   
min               0.36         -7.62          14.01  -104.04   
25%               2.10         -1.68         140.05     1.79   
50%               2.36         -1.10         194.12     3.26   
75%               2.52         -0.57         320.12     5.77   
max               3.38          0.83        1758.85  1512.24   

      MaxAbsEStateIndex MinEStateIndex NumHeteroatoms PEOE_VSA10 PEOE_VSA14  \
count           9893.00        9893.00        9893.00    9893.00    9893.00   
mean              10.56          -2.10           8.59      11.03      18.49   
std                1.57           2.06           7.66      13.91      35.83   
min                0.00          -6.12           0.00       0.00       0.00   
25%                9.95          -4.64           4.00       0.00       0.00   
50%               10.42          -1.25           6.00       6.04       5.97   
75%               11.53          -0.79          10.00      18.31      17.91   
max               14.63           4.75          53.00     122.02     482.43   

      PEOE_VSA6 PEOE_VSA7 PEOE_VSA8 SMR_VSA10 SMR_VSA5 SlogP_VSA3 VSA_EState9  \
count   9893.00   9893.00   9893.00   9893.00  9893.00    9893.00     9893.00   
mean       8.92     11.22      6.79     15.81    30.92      13.60       49.43   
std       19.36     20.02     10.78     18.27    33.53      14.69       29.54   
min        0.00      0.00      0.00      0.00     0.00       0.00       -5.06   
25%        0.00      0.00      0.00      5.97     6.42       4.79       30.00   
50%        0.00      5.56      0.00     11.75    19.77       9.59       41.67   
75%       12.13     13.34      6.92     17.74    42.73      14.38       56.08   
max      258.84    271.42     96.92     79.92   492.73     115.41      384.45   

        fr_COO  fr_COO2  
count  9893.00  9893.00  
mean      0.46     0.46  
std       0.68     0.68  
min       0.00     0.00  
25%       0.00     0.00  
50%       0.00     0.00  
75%       1.00     1.00  
max       8.00     8.00

BertzCT     Chi1    Chi1n    Chi1v    Chi2n    Chi2v    Chi3v    Chi4n  \
count  1039.00  1039.00  1039.00  1039.00  1039.00  1039.00  1039.00  1039.00   
mean    600.90    10.48     6.88     7.97     5.28     6.30     4.18     2.28   
std     657.37     8.82     6.30     7.41     5.04     6.12     4.34     2.61   
min       0.00     0.00     0.00     0.00     0.00     0.00     0.00     0.00   
25%     150.19     4.69     2.84     3.08     1.98     2.19     1.22     0.52   
50%     315.70     7.22     4.38     5.24     3.29     3.96     2.37     1.21   
75%     817.83    13.67     9.42    10.28     7.42     8.61     6.14     3.34   
max    4069.96    73.27    58.85    61.91    43.05    45.43    30.33    20.41   

      EState_VSA1 EState_VSA2 ExactMolWt FpDensityMorgan1 FpDensityMorgan2  \
count     1039.00     1039.00    1039.00          1039.00          1039.00   
mean        33.72       12.67     333.93             0.61             1.16   
std         39.50       16.44     282.25            20.70            20.72   
min          0.00        0.00       1.01          -666.00          -666.00   
25%          5.97        0.00     149.05             1.00             1.55   
50%         18.12        6.42     227.09             1.20             1.82   
75%         46.29       17.25     425.59             1.49             2.07   
max        363.71       99.94    2237.34             3.00             3.20   

      FpDensityMorgan3 HallKierAlpha HeavyAtomMolWt   Kappa3  \
count          1039.00       1039.00        1039.00  1039.00   
mean              1.59         -1.35         313.87    10.10   
std              20.74          1.22         264.81    71.41   
min            -666.00         -7.73           0.00  -104.04   
25%               2.00         -1.85         140.05     1.92   
50%               2.31         -1.10         215.03     3.62   
75%               2.54         -0.53         399.69     7.23   
max               3.40          0.96        2035.13  1512.24   

      MaxAbsEStateIndex MinEStateIndex NumHeteroatoms PEOE_VSA10 PEOE_VSA14  \
count           1039.00        1039.00        1039.00    1039.00    1039.00   
mean              10.50          -2.13           9.46      12.80      23.90   
std                2.09           2.15           8.68      16.44      47.73   
min                0.00          -6.43           0.00       0.00       0.00   
25%                9.87          -4.60           4.00       0.00       0.00   
50%               10.50          -1.32           6.00       6.04       7.82   
75%               11.84          -0.70          12.00      22.91      23.47   
max               15.81           6.00          53.00     122.02     482.43   

      PEOE_VSA6 PEOE_VSA7 PEOE_VSA8 SMR_VSA10 SMR_VSA5 SlogP_VSA3 VSA_EState9  \
count   1039.00   1039.00   1039.00   1039.00  1039.00    1039.00     1039.00   
mean      14.45     15.72      8.62     18.59    39.54      15.61       54.30   
std       30.86     27.38     13.43     20.52    50.79      17.24       40.32   
min        0.00      0.00      0.00      0.00     0.00       0.00      -10.03   
25%        0.00      0.00      0.00      5.97     6.61       4.79       29.33   
50%        0.00      6.08      0.00     11.94    24.54       9.59       42.92   
75%       18.20     18.73     12.26     21.43    57.47      19.18       69.74   
max      375.43    271.42    100.35     95.82   624.27     115.41      384.45   

        fr_COO  fr_COO2      EC1      EC2      EC3      EC4      EC5      EC6  
count  1039.00  1039.00  1039.00  1039.00  1039.00  1039.00  1039.00  1039.00  
mean      0.48     0.48     0.54     0.59     0.30     0.24     0.13     0.13  
std       0.82     0.82     0.50     0.49     0.46     0.42     0.33     0.33  
min       0.00     0.00     0.00     0.00     0.00     0.00     0.00     0.00  
25%       0.00     0.00     0.00     0.00     0.00     0.00     0.00     0.00  
50%       0.00     0.00     1.00     1.00     0.00     0.00     0.00     0.

BertzCT      Chi1     Chi1n     Chi1v     Chi2n     Chi2v     Chi3v  \
count  15859.00  15859.00  15859.00  15859.00  15859.00  15859.00  15859.00   
mean     520.79      9.22      5.92      6.82      4.49      5.32      3.47   
std      551.12      6.98      4.78      5.99      3.86      5.02      3.51   
min        0.00      0.00      0.00      0.00      0.00      0.00      0.00   
25%      149.10      4.68      2.84      2.93      1.95      2.04      1.16   
50%      292.46      6.51      4.07      4.43      2.99      3.29      1.97   
75%      653.44     11.27      7.60      8.61      5.83      6.75      4.63   
max     4069.96     73.27     58.85     61.91     43.05     45.43     30.33   

          Chi4n EState_VSA1 EState_VSA2 ExactMolWt FpDensityMorgan1  \
count  15859.00    15859.00    15859.00   15859.00         15859.00   
mean       1.81       29.48       10.59     295.30             1.20   
std        1.93       32.31       13.86     229.77             7.50   
min        0.00        0.00        0.00       1.01          -666.00   
25%        0.50        5.97        0.00     148.04             1.04   
50%        1.08       17.47        6.42     206.13             1.25   
75%        2.57       44.88       12.97     348.05             1.50   
max       20.41      363.71       99.94    2237.34             3.00   

      FpDensityMorgan2 FpDensityMorgan3 HallKierAlpha HeavyAtomMolWt  \
count         15859.00         15859.00      15859.00       15859.00   
mean              1.77             2.21         -1.22         277.47   
std               7.51             7.52          0.96         216.70   
min            -666.00          -666.00         -7.73           0.00   
25%               1.68             2.09         -1.68         136.11   
50%               1.86             2.36         -1.10         196.12   
75%               2.06             2.50         -0.57         328.24   
max               3.20             3.40          0.96        2035.13   

         Kappa3 MaxAbsEStateIndex MinEStateIndex NumHeteroatoms PEOE_VSA10  \
count  15859.00          15859.00       15859.00       15859.00   15859.00   
mean       6.15             10.55          -2.12           8.64      11.13   
std       47.87              1.60           2.07           7.72      14.14   
min     -104.04              0.00          -6.43           0.00       0.00   
25%        1.79              9.92          -4.66           4.00       0.00   
50%        3.27             10.42          -1.27           6.00       6.04   
75%        5.87             11.56          -0.79          10.00      18.31   
max     1512.24             15.81           6.00          53.00     122.02   

      PEOE_VSA14 PEOE_VSA6 PEOE_VSA7 PEOE_VSA8 SMR_VSA10  SMR_VSA5 SlogP_VSA3  \
count   15859.00  15859.00  15859.00  15859.00  15859.00  15859.00   15859.00   
mean       18.19      9.32     11.61      6.83     15.86     31.61      13.76   
std        35.61     20.70     20.75     11.06     18.27     35.32      14.79   
min         0.00      0.00      0.00      0.00      0.00      0.00       0.00   
25%         0.00      0.00      0.00      0.00      5.97      6.42       4.79   
50%         5.97      0.00      0.00      0.00     11.75     24.10       9.59   
75%        16.37     12.13     13.85      6.92     17.72     45.13      15.78   
max       482.43    375.43    271.42    100.35     95.82    624.27     115.41   

      VSA_EState9    fr_COO   fr_COO2       EC1       EC2       EC3       EC4  \
count    15859.00  15859.00  15859.00  15859.00  15859.00  15859.00  15859.00   
mean        49.63      0.46      0.46      0.66      0.79      0.31      0.28   
std         30.06      0.68      0.68      0.47      0.41      0.46      0.45   
min        -10.03      0.00      0.00      0.00      0.00      0.00      0.00   
25%         30.00      0.00      0.00      0.00      1.00      0.00      0.00   
50%         41.67      0.00      0.00      1.00      1.00      0.00      0.00   
75%    

(Index(['BertzCT', 'Chi1', 'Chi1n', 'Chi1v', 'Chi2n', 'Chi2v', 'Chi3v', 'Chi4n',
        'EState_VSA1', 'EState_VSA2', 'ExactMolWt', 'FpDensityMorgan1',
        'FpDensityMorgan2', 'FpDensityMorgan3', 'HallKierAlpha',
        'HeavyAtomMolWt', 'Kappa3', 'MaxAbsEStateIndex', 'MinEStateIndex',
        'NumHeteroatoms', 'PEOE_VSA10', 'PEOE_VSA14', 'PEOE_VSA6', 'PEOE_VSA7',
        'PEOE_VSA8', 'SMR_VSA10', 'SMR_VSA5', 'SlogP_VSA3', 'VSA_EState9',
        'fr_COO', 'fr_COO2', 'EC1', 'EC2', 'EC3', 'EC4', 'EC5', 'EC6'],
       dtype='object'),
 Index(['BertzCT', 'Chi1', 'Chi1n', 'Chi1v', 'Chi2n', 'Chi2v', 'Chi3v', 'Chi4n',
        'EState_VSA1', 'EState_VSA2', 'ExactMolWt', 'FpDensityMorgan1',
        'FpDensityMorgan2', 'FpDensityMorgan3', 'HallKierAlpha',
        'HeavyAtomMolWt', 'Kappa3', 'MaxAbsEStateIndex', 'MinEStateIndex',
        'NumHeteroatoms', 'PEOE_VSA10', 'PEOE_VSA14', 'PEOE_VSA6', 'PEOE_VSA7',
        'PEOE_VSA8', 'SMR_VSA10', 'SMR_VSA5', 'SlogP_VSA3', 'VSA_EState9',
        

In [4]:
# Check for number of zeros in the data to see if they are meaningful

zero_counts_train = (train == 0).sum()
zero_counts_test = (test == 0).sum()
zero_counts_original = (original == 0).sum()
zero_counts_train1 = (train1 == 0).sum()

zero_sum = pd.concat([zero_counts_train, zero_counts_test, zero_counts_original, zero_counts_train1],axis=1)
zero_sum = zero_sum.rename(columns={0:'train',1:'test',2:'original',3:'train1'})
zero_sum

train    test  original  train1
BertzCT               34    23.0         8      42
Chi1                  36    21.0         6      42
Chi1n                110    87.0        14     124
Chi1v                109    87.0        14     123
Chi2n                229   160.0        25     254
Chi2v                225   160.0        25     250
Chi3v                493   306.0        48     541
Chi4n               1310   847.0        87    1396
EState_VSA1         3010  1995.0       204    3212
EState_VSA2         5731  3794.0       372    6091
ExactMolWt             0     0.0         0       0
FpDensityMorgan1       1     0.0         0       1
FpDensityMorgan2       1     0.0         0       1
FpDensityMorgan3       1     0.0         0       1
HallKierAlpha         16    12.0         3      19
HeavyAtomMolWt         1     0.0         1       2
Kappa3                 2     0.0         1       3
MaxAbsEStateIndex     24    22.0         6      30
MinEStateIndex        27    23.0         6      33
NumHeteroatoms        41    28.0         9      50
PEOE_VSA10          6205  4107.0       418    6621
PEOE_VSA14          4181  2778.0       288    4463
PEOE_VSA6           9985  6679.0       630   10600
PEOE_VSA7           7485  4917.0       479    7950
PEOE_VSA8           8473  5568.0       542    9002
SMR_VSA10           2406  1615.0       162    2562
SMR_VSA5            2239  1450.0       163    2402
SlogP_VSA3          2595  1731.0       185    2778
VSA_EState9           63    63.0        13      76
fr_COO              9325  6246.0       673    9984
fr_COO2             9311  6234.0       668    9965
EC1                 4930     NaN       476    5400
EC2                 2983     NaN       428    3406
EC3                10182     NaN       732   10900
EC4                10697     NaN       794   11476
EC5                12689     NaN       907   13581
EC6                12589     NaN       908   13479

In [5]:
train.isna().any().any(),test.isna().any().any(),original.isna().any().any(),train1.isna().any().any(),

(False, False, False, False)

No missing data.

# Check for duplicated rows

In [6]:
features = [f for f in test.columns if not (f.startswith('EC') or f=='id')]
features1 = features + ['EC1', 'EC2']

dup_train = train.duplicated()
print(f'Duplicated number of rows in train set (features columns + EC1~EC6 columns) = {dup_train.sum()}')
dup_train = train[features].duplicated()
print(f'Duplicated number of rows in train set (features columns only) = {dup_train.sum()}')
dup_train = train[features1].duplicated()
print(f'Duplicated number of rows in train set (features columns + EC1~EC2 columns) = {dup_train.sum()}')
print("\n")

dup_test = test.duplicated()
print(f'Duplicated number of rows in test set (features columns only) = {dup_test.sum()}')
print("\n")

dup_original = original.duplicated()
print(f'Duplicated number of rows in original set (features columns + EC1~EC6 columns) = {dup_original.sum()}')
dup_original = original[features].duplicated()
print(f'Duplicated number of rows in original set (features columns only) = {dup_original.sum()}')
dup_original = original[features1].duplicated()
print(f'Duplicated number of rows in original set (features columns + EC1~EC2 columns) = {dup_original.sum()}')
print("\n")

dup_train1 = train1.duplicated()
print(f'Duplicated number of rows in train1 set (features columns + EC1~EC6 columns) = {dup_train1.sum()}')
dup_train1 = train1[features].duplicated()
print(f'Duplicated number of rows in train1 set (features columns only)= {dup_train1.sum()}')
dup_train1 = train1[features1].duplicated()
print(f'Duplicated number of rows in train1 set (features columns + EC1~EC2 columns) = {dup_train1.sum()}')

Duplicated number of rows in train set (features columns + EC1~EC6 columns) = 0
Duplicated number of rows in train set (features columns only) = 0
Duplicated number of rows in train set (features columns + EC1~EC2 columns) = 0


Duplicated number of rows in test set (features columns only) = 0


Duplicated number of rows in original set (features columns + EC1~EC6 columns) = 18
Duplicated number of rows in original set (features columns only) = 95
Duplicated number of rows in original set (features columns + EC1~EC2 columns) = 46


Duplicated number of rows in train1 set (features columns + EC1~EC6 columns) = 0
Duplicated number of rows in train1 set (features columns only)= 90
Duplicated number of rows in train1 set (features columns + EC1~EC2 columns) = 33


In [7]:
dup_train_test = pd.concat([train[features],test[features]],axis=0).duplicated()
print(f'Duplicated number of rows between train and test sets (features columns only) = {dup_train_test.sum()}')
dup_train1_test = pd.concat([train1[features],test[features]],axis=0).duplicated()
print(f'Duplicated number of rows between train1 and test sets (features columns only) = {dup_train1_test.sum()}')
dup_ori_test = pd.concat([original[features],test[features]],axis=0).duplicated()
print(f'Duplicated number of rows between original and test sets (features columns only) = {dup_ori_test.sum()}')
print("\n")

dup_train_ori = pd.concat([train,original],axis=0).duplicated()
print(f'Duplicated number of rows between train and original sets (features columns + EC1~EC6 columns) = {dup_train_ori.sum()}')
dup_train_ori = pd.concat([train[features],original[features]],axis=0).duplicated()
print(f'Duplicated number of rows between train and original sets (features columns only) = {dup_train_ori.sum()}')
dup_train_ori = pd.concat([train[features1],original[features1]],axis=0).duplicated()
print(f'Duplicated number of rows between train and original sets (features columns + EC1~EC2 columns) = {dup_train_ori.sum()}')

Duplicated number of rows between train and test sets (features columns only) = 0
Duplicated number of rows between train1 and test sets (features columns only) = 100
Duplicated number of rows between original and test sets (features columns only) = 105


Duplicated number of rows between train and original sets (features columns + EC1~EC6 columns) = 18
Duplicated number of rows between train and original sets (features columns only) = 108
Duplicated number of rows between train and original sets (features columns + EC1~EC2 columns) = 51


Notes:
- 14838 (+1039 original data) rows for training, 9893 rows for prediction.
- No missing data in all rows.
- Excluding id, 37 columns in train set and 31 columns in test set.
- All features are numerical.
- "NumHeteroatoms", "fr_COO" and "fr_COO2" are integers.
- "EC1" to "EC6" are binary.